File: 2487-remove-nodes-from-linked-list/2487-remove-nodes-from-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNodes(ListNode head) {
        Stack<ListNode> st = new Stack<>();
        while(head != null) {
            while(!st.isEmpty() && st.peek().val < head.val) {
                st.pop();
            }
            st.add(head);
            head = head.next;
        }

        ListNode prev = null;
        while(!st.isEmpty()) {
            ListNode curr = st.pop();
            curr.next = prev;
            prev = curr;
        }

        return prev;
    }
}

File: 2487-remove-nodes-from-linked-list/2487-remove-nodes-from-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var removeNodes = function(head) {
    let st = [];
    while(head != null) {
        while(st.length > 0 && st[st.length - 1].val < head.val) {
            st.pop();
        }
        st.push(head);
        head = head.next;
    }
    
    let prev = null;
    while(st.length > 0) {
        let curr = st[st.length - 1];
        st.pop();
        curr.next = prev;
        prev = curr;
    }
    
    return prev;
};